### PR TITLE
Rational Bezier Splines

### PR DESCRIPTION
**Pull request by jzwar**
### Rational Bezier Splines
The main feature added in this PR are rational Bezier splines from BezMan. Other changes include:
1. Operations on Rational Bezier splines
2. Compositions between arbitray combinations of rational and polynomial Bezier splines
3. `_check_an_update_c` method, that makes initializing splines safer and also prevents a lot of unneccessary segmentation faults
4. Bezier patch extraction for BSplines and NURBS, which can be used to extract patches and perform compositions on more complex geometries

**Some changes are still required, which in particular are related to BezierGroups, which are currently changed to use dynamic polymorphism. If these changes are completed, bezman (just like Splinelib) will be included as a submodule to splinepy**

## Operations on Rational Bezier Splines
- Initializiation of Rational bezier splines
- multiplication with a scalar
- multiplication with one another (no support for polynomial splines yet)
- addition with point type and or rational bezier spline
## Compositions
Compositions are now supported for any combination of Bezier type splines. The functions are passed to the c-objects accordingly which makes these operations much cheaper (as compared to transforming Polynomial splines into rational splines prior to composition)
## Sanity checks
Up until now, backend objects were created, regardless of whether the data is coherent or not. Now, whenever the backend object is updated, the coherence of the underlying data is checked and errors are raised otherwise in order to prevent segmentation faults or inconsistent data
## Bezier patch extraction
Individual knot spans, are extracted by means of knot insertion. This is done using functions of the helper file. Further type identifier were introduced to facilitate the type identification in template functions.

### CODE REVIEW

**Comment by jzwar** [comment]
Not all files have been cleaned up. This is done after this PR is merged to prevent to many changes at once. Only the once I worked on were autoformated

**Review by j042 — COMMENTED** [review]
Nice updates!
I need a bit more time to review python implementations, but it is on it way

File: cpp/splinepy/py/py_bezier.hpp [hunk 1]
```diff
@@ -39,64 +38,58 @@ class PyBezier {
   Bezier<para_dim, dim> c_bezier;
 
   const std::string whatami =
-      "Bezier, parametric dimension: "
-      + std::to_string(para_dim)
-      + ", physical dimension: "
-      + std::to_string(dim);
+      "Bezier, parametric dimension: " + std::to_string(para_dim) +
+      ", physical dimension: " + std::to_string(dim);
```
[the remark this answers is not in the document]
Comment (jzwar) [reply cpp/splinepy/py/py_bezier.hpp:42]: No thanks. I prefer the traditional c++ style, as it is clear that something follows in the next line

File: cpp/splinepy/py/py_bspline.hpp [hunk 2]
```diff
@@ -603,74 +580,49 @@ class PyBSpline {
     update_c();
 
     return residual;
-
   }
 
-  void interpolate_curve(py::array_t<double> points,
-                         int degree,
+  void interpolate_curve(py::array_t<double> points, int degree,
```
[the remark this answers is not in the document]
Comment (jzwar) [reply cpp/splinepy/py/py_bspline.hpp:585]: Next PR. I wrote issue #24

File: cpp/splinepy/py/py_nurbs.hpp [hunk 3]
```diff
@@ -1,35 +1,36 @@
-#include <memory>
```
[the remark this answers is not in the document]
Comment (jzwar) [reply cpp/splinepy/py/py_nurbs.hpp:1]: It orders everything alphabetically

File: cpp/splinepy/py/py_rational_bezier.hpp [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    std::conditional_t<
+      (dim > 1),
+      bezman::Point<static_cast<unsigned>(dim)>,
+      double>,
+    double
+>;
+
+
+template<int para_dim, int dim>
+class PyRationalBezier {
```
[the remark this answers is not in the document]
Comment (jzwar) [reply cpp/splinepy/py/py_rational_bezier.hpp:27]: nope

File: cpp/splinepy/splines/helpers.hpp [hunk 5]
```diff
@@ -63,5 +59,152 @@ GetParametricBounds(const SplineType& spline) {
...
+          // CTPS non-weighted
+          ctps,
+          // weights
+          weights));
+    } else {
+      bezier_list.push_back(ReturnType(
+          // degrees
+          degrees,
+          // ctps
+          ctps));
```
[the remark this answers is not in the document]
Comment (jzwar) [reply cpp/splinepy/splines/helpers.hpp:205]: `c`haotic, `c`lumsy, `c`razy if you use it

File: splinepy/_spline.py [hunk 6]
```diff
@@ -342,12 +343,11 @@ def unique_knots(self,):
         """
         logging.debug("Spline - Computing unique knots using `np.unique`.")
         unique_knots = []
-        if self.whatami.startswith("Bezier"):
```
[the remark this answers is not in the document]
Comment (jzwar) [reply splinepy/_spline.py:345]: Because RationalBezier does not start with Bezier <3

File: cpp/splinepy/splines/helpers.hpp [hunk 7]
```diff
@@ -63,5 +59,155 @@ GetParametricBounds(const SplineType& spline) {
...
-} /* namespace splinepy::splines */
+  // Init return value
+  std::vector<ReturnType> bezier_list{};
+  bezier_list.reserve(n_total_patches);
+
+  // Loop over the individual patches
+  for (int i_patch{}; i_patch < n_total_patches; i_patch++) {
+    // Determine internal postitions in local coord system
+    std::array<std::size_t, para_dim> patch_ctp_id_offsets{};
+    int ii{i_patch};
```
Comment (j042) [inline cpp/splinepy/splines/helpers.hpp:128]: REAL NICE

File: cpp/splinepy/py/py_rational_bezier.hpp [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+           &PyRationalBezier<para_dim, dim>::template ComposeRP<1>,
+           py::arg("inner_function"))
+      // Composition with a surface
+      .def("compose_surface_rp",
+           &PyRationalBezier<para_dim, dim>::template ComposeRP<2>,
+           py::arg("inner_function"))
+      // Composition with a volume
+      .def("compose_volume_rp",
+           &PyRationalBezier<para_dim, dim>::template ComposeRP<3>,
+           py::arg("inner_function"))
```
Comment (j042) [inline cpp/splinepy/py/py_rational_bezier.hpp:332]: Special names will make other beziers jealous

File: splinepy/bezier.py [hunk 9]
```diff
@@ -255,17 +254,17 @@ def compose(self, inner_function):
...
-                res_c_spl = self._c_spline.compose_surface(
+                res_c_spl = self._c_spline.compose_surface_pp(
                     inner_function._c_spline
                 )
 
             elif inner_function.para_dim == 3:
-                res_c_spl = self._c_spline.compose_volume(
+                res_c_spl = self._c_spline.compose_volume_pp(
                     inner_function._c_spline
                 )
```
Comment (j042) [inline splinepy/bezier.py:269]: as discussed, this should be replaced to be N-dim compatible

File: splinepy/bezier.py [hunk 10]
```diff
@@ -276,50 +275,39 @@ def compose(self, inner_function):
...
+            elif inner_function.para_dim == 3:
+                res_c_spl = self._c_spline.compose_volume_pr(
+                    inner_function._c_spline
+                )
 
-        Parameters
-        -----------
-        tolist : bool
-          Default is False. Convert numpy properties into lists
+            else :
```
Comment (j042) [inline splinepy/bezier.py:299]: N-dim

File: splinepy/nurbs.py [hunk 11]
```diff
@@ -155,50 +101,28 @@ def _update_p(self,):
...
+
+        # Transform into Rational Bezier Splinepy objects
+        extracted_bezier_patches = []
+        for c_object_spline in list_of_c_object_beziers:
+            i_rational_bezier = RationalBezier()
+            i_rational_bezier._c_spline = c_object_spline
+            i_rational_bezier._update_p()
+            extracted_bezier_patches.append(i_rational_bezier)
+
+        return extracted_bezier_patches
```
Comment (j042) [inline splinepy/nurbs.py:128]: Just as a reminder for discussion: would it be nice to have an intermediate abstractclass?

File: splinepy/rational_bezier.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+        )
+        return self.evaluate(q)
+
+    def __mul__(self, factor):
+        """
+        Overloads multiplication between splines with different types of
+        degrees
+
+        The resulting spline fulfills the equation
+          factora(t) * factorb(t) = result(t)
```
Comment (j042) [inline splinepy/rational_bezier.py:134]: does this mean 
`factor_a.evaluate(t) * factor_b.evaluate(t) = result.evaluate(t)`?

File: splinepy/rational_bezier.py [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+        else:
+            raise TypeError("Multiplication with unknown type requested.")
+
+    def __add__(self, summand):
+        """
+        Calculates the spline that formes the sum of the summand and the
+        current spline (function argument)
+
+        The resulting spline fulfils the equation
+          self(t) + summand(t) = result(t)
```
Comment (j042) [inline splinepy/rational_bezier.py:185]: +`.evaluate()`?
In that case, I guess I will have to retract my comment above

File: splinepy/rational_bezier.py [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+                "Sum must be formed between RationalBezier Splines")
+
+    def compose(self, inner_function):
+        """
+        Calculates the spline that formes the composition of the inner function
+        spline (function argument), using the caller spline as the outer (or
+        deformation function).
+
+        The resulting spline fulfils the equation
+          spline_self(inner_function(t)) = result(t)
```
Comment (j042) [inline splinepy/rational_bezier.py:227]: `.evaluate()`?

File: splinepy/rational_bezier.py [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+                    inner_function._c_spline
+                )
+
+            elif inner_function.para_dim == 3:
+                res_c_spl = self._c_spline.compose_volume_rr(
+                    inner_function._c_spline
+                )
+
+            else:
+                raise TypeError(
```
Comment (j042) [inline splinepy/rational_bezier.py:259]: more! I know you can

**Review by j042 — CHANGES_REQUESTED** [review]
v2

File: splinepy/rational_bezier.py (hunk 12 above)
Comment (jzwar) [reply splinepy/rational_bezier.py:134]: yes

File: cpp/splinepy/py/py_bezier.hpp [hunk 16]
```diff
@@ -226,90 +220,84 @@ class PyBezier {
...
   // Composition Routine
-  template<int par_dim_inner_function>
-  PyBezier<par_dim_inner_function, dim> Compose(
-              const PyBezier<par_dim_inner_function, para_dim>& inner_function){
-      // Use Composition routine
-      PyBezier<par_dim_inner_function, dim> result{(*this).c_bezier.Compose(
-                        inner_function.c_bezier)
-                        };
+  template <int par_dim_inner_function>
+  PyBezier<par_dim_inner_function, dim> ComposePP(
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/py/py_bezier.hpp:224]: next PR

File: cpp/splinepy/py/py_bezier.hpp [hunk 17]
```diff
@@ -226,90 +220,84 @@ class PyBezier {
...
+    PyBezier<par_dim_inner_function, dim> result{
+        (*this).c_bezier.Compose(inner_function.c_bezier)};
     result.update_p();
     return result;
   }
 
-
+  // Composition Routine with Rational Spline
+  template <int par_dim_inner_function>
+  PyRationalBezier<par_dim_inner_function, dim> ComposePR(
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/py/py_bezier.hpp:235]: next PR

File: cpp/splinepy/py/py_rational_bezier.hpp [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+  template <int par_dim_inner_function>
+  PyRationalBezier<par_dim_inner_function, dim> ComposeRP(
+      const PyBezier<par_dim_inner_function, para_dim>& inner_function) {
+    // Use Composition routine
+    PyRationalBezier<par_dim_inner_function, dim> result{
+        (*this).c_rational_bezier.Compose(inner_function.c_bezier)};
+    result.update_p();
+    return result;
+  }
+};
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/py/py_rational_bezier.hpp:256]: next PR

File: cpp/splinepy/py/py_bezier.hpp [hunk 19]
```diff
@@ -226,90 +220,84 @@ class PyBezier {
...
+           py::arg("inner_function"))
+      .def("compose_volume_pp", &PyBezier<para_dim, dim>::template ComposePP<3>,
+           py::arg("inner_function"))
+      .def("compose_line_pr", &PyBezier<para_dim, dim>::template ComposePR<1>,
+           py::arg("inner_function"))
+      .def("compose_surface_pr",
+           &PyBezier<para_dim, dim>::template ComposePR<2>,
+           py::arg("inner_function"))
+      .def("compose_volume_pr", &PyBezier<para_dim, dim>::template ComposePR<3>,
+           py::arg("inner_function"))
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/py/py_bezier.hpp:288]: next PR

File: cpp/splinepy/py/py_bezier.hpp [hunk 20]
```diff
@@ -197,27 +192,26 @@ class PyBezier {
...
-  PyBezier add_spline (const PyBezier& a){
+  PyBezier add_spline(const PyBezier& a) {
     PyBezier result{(*this).c_bezier + a.c_bezier};
     result.update_p();
     return result;
   }
-  
-  PyBezier multiply_with_scalar_spline (const PyBezier<para_dim,1>& a){
+
+  PyBezier multiply_with_scalar_spline(const PyBezier<para_dim, 1>& a) {
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/py/py_bezier.hpp:214]: next PR

File: cpp/splinepy/py/py_rational_bezier.hpp (hunk 8 above)
Comment (j042) [reply cpp/splinepy/py/py_rational_bezier.hpp:332]: next PR

File: splinepy/bezier.py (hunk 9 above)
Comment (j042) [reply splinepy/bezier.py:269]: next PR

File: splinepy/bezier.py (hunk 10 above)
Comment (j042) [reply splinepy/bezier.py:299]: next PR

File: cpp/splinepy/py/py_bezier.hpp [hunk 21]
```diff
@@ -197,27 +192,26 @@ class PyBezier {
...
-    );
+        static_cast<std::size_t>(p_dim));
 
     update_p();
   }
 
   // Multiplication routines
-  PyBezier<para_dim,1> multiply_with_spline (const PyBezier& a){
-    PyBezier<para_dim,1> result{(*this).c_bezier * a.c_bezier};
+  PyBezier<para_dim, 1> multiply_with_spline(const PyBezier& a) {
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/py/py_bezier.hpp:201]: next PR

**Review by j042 — APPROVED** [review]
As discussed, I will merge this one and leave some TODOs for next PR: n-dim bezier operations and some minor renaming.  
Good stuff @jzwar!

**Pull request merged**